draft-ietf-avtcore-rtp-over-quic.md

```diff
@@ -860,41 +860,33 @@ following Payload-specific RTP Feedback (PSFB) feedback.
 
-RTP itself does not specify a congestion control algorithm, but {{!RFC8888}} defines an RTCP
-feedback message intended to enable rate adaptation for interactive real-time traffic using RTP,
-and successful rate adaptation will accomoplish congestion control as well.
-Various rate adaptation algorithms for real-time media are defined in separate RFCs
-(e.g. SCReAM {{!RFC8298}} and NADA {{!RFC8698}}).
-The rate adaptation algorithms for RTP are specifically tailored for
-real-time transmissions at low latencies. The available rate adaptation
-algorithms for RTP expose a `target_bitrate` that can be used to dynamically
-reconfigure media codecs to produce media at a rate that can be sent in
-real-time under the observed network conditions.
+This sectioon also gives guidance about avoiding problems with "nested" congestion controllers, in {{nested-CC}}.
 
-This section defines two architectures for congestion control and bandwidth
-estimation for RoQ, but it does not mandate any specific rate adaptation algorithm
-to use. The section also discusses congestion control
+This section also discusses congestion control
 implications of using shared or multiple separate QUIC connections to send and
-receive multiple independent data streams.
+receive multiple independent data streams, in {{shared-connections}}.
 
 It is assumed that the congestion controller in use provides a pacing mechanism
 to determine when a packet can be sent to avoid bursts. The currently proposed
-congestion control algorithms for real-time communications provide such pacing
+congestion control algorithms for real-time communications (e.g. SCReAM and NADA) provide such pacing
 mechanisms. The use of congestion controllers which don't provide a pacing
 mechanism is out of scope of this document.
 
-## Congestion Control at the QUIC layer {#cc-quic-layer}
+## Congestion Control at the Transport Layer {#cc-quic-layer}
+
+QUIC is a congestion controlled transport protocol. Senders are required to
+employ some form of congestion control. The default congestion control specified
+for QUIC in {{!RFC9002}} is similar to TCP NewReno {{?RFC6582}}, but senders are free to choose
+any congestion control algorithm as long as they follow the guidelines specified
+in {{Section 3 of ?RFC8085}}, and QUIC implementors make use of this freedom.
 
 If congestion control is to be applied at the transport layer, it is RECOMMENDED
 that the QUIC Implementation uses a congestion controller that keeps queueing
 delays short to keep the transmission latency for RTP and RTCP packets as low
-as possible.
+as possible. One example of a congestion controller that does this is "Google Congestion Controller" {{!I-D.draft-ietf-rmcat-gcc}}.
```

Comment:
This is a reasonable question. I think the answer is that GCC targets transport protocols that are not aware that they are carrying media, while NADA and SCReAM target RTP applications that are aware they are performing rate adaptation for media, so the thing to do in this transport protocol section, is to cite GCC.

Does that make sense?

Reply:
To me, all three seem to target RTP applications, and the GCC draft references multiple RTP header extensions and RTCP feedback messages which were explicitly designed for GCC.

Another alternative to mention could be BBR. However, I don't know if that is a good recommendation for media. I think it has been mentioned somewhere that BBRs probing phases are problematic for real-time media.

BBR also (still?) seems to be a draft, so we might not want to use it as a normative reference, either. I think we have this problem with other references, too, but I'd like to avoid adding more normative references to draft documents if possible (at least if the documents are expired and don't seem to be actively developed any longer). Regardless of which algorithm we reference here, does the reference have to be normative?

Reply:
@mengelbart - I'm replying here, just to preserve our helpful discussion which was also happening in Slack. Here's what I've got, point by point. 

> To me, all three seem to target RTP applications, and the GCC draft references multiple RTP header extensions and RTCP feedback messages which were explicitly designed for GCC.

I think we're struggling a bit, because the draft is not internally consistent (yet) - for example, if we are writing a specification that assumes QUIC implementations know that our packets are media (perhaps because the connection uses rtp-mux-quic as its ALPN, rather than h3, or some other non-media ALPN), that's different from assuming that QUIC implementations don't know that our packets are media, and have to use a congestion controller-rate adapter that works for any packets, without special knowledge. 

As I said in SLACK, if we CAN have real-time and non-real-time packets in the same QUIC connection, that has different implications. So, we need to be looking for inconsistencies. 

> Another alternative to mention could be BBR. However, I don't know if that is a good recommendation for media. I think it has been mentioned somewhere that BBRs probing phases are problematic for real-time media.

Yes, exactly. The idea is that any probing congestion controller probes by sending "more data than it did the last RTT", until it detects increasing delay, or actual packet loss. If the media congestion controller doesn't need to probe for bandwidth, or doesn't do that often, the media traffic won't experience increasing delays or packet loss. 

> BBR also (still?) seems to be a draft, so we might not want to use it as a normative reference, either. I think we have this problem with other references, too, but I'd like to avoid adding more normative references to draft documents if possible (at least if the documents are expired and don't seem to be actively developed any longer). Regardless of which algorithm we reference here, does the reference have to be normative?

Right. So the point for normative references is that people must have access to them, to implement YOUR specification. What I THINK we are doing, is to say "use a media-centric rate adaptation mechanism, and here are three that could meet your needs". IMO, all three of these are informational, not normative (because you don't have to read any of them, you just need to understand what makes a rate adaptation mechanism media-centric). 

If you think it would help, we can add [RFC 8836 Congestion Control Requirements for Interactive Real-Time Media](https://www.rfc-editor.org/rfc/rfc8836.html). I'm still not convinced that would be normative, but it might be helpful to the reader. 

Got any thoughts about any of this? I'll be typing, but I'm happy to be corrected, especially while I haven't merged any of this into Main yet!

Reply:
That all makes sense. 

> I think we're struggling a bit, because the draft is not internally consistent (yet) - for example, if we are writing a specification that assumes QUIC implementations know that our packets are media (perhaps because the connection uses rtp-mux-quic as its ALPN, rather than h3, or some other non-media ALPN), that's different from assuming that QUIC implementations don't know that our packets are media, and have to use a congestion controller-rate adapter that works for any packets, without special knowledge.

I think that's why we use `RECOMMEND` above. We don't know if the QUIC stack is aware of what is being sent over it and if the application has any control over the congestion control used in the QUIC stack. If the app is in control, we know what is being sent and can decide to use real-time CC. We can't do much if neither the app can choose a congestion controller nor the QUIC stack knows about the payload.

> As I said in SLACK, if we CAN have real-time and non-real-time packets in the same QUIC connection, that has different implications. So, we need to be looking for inconsistencies.

Agree. I think we should add more clarification to section 7.4 about the implications for congestion control when sharing a connection between real-time and non-real-time streams.

> IMO, all three of these are informational, not normative (because you don't have to read any of them, you just need to understand what makes a rate adaptation mechanism media-centric).

Having an informational reference to any of the three algorithms should be fine, I think :) And adding a reference to RFC 8836 also sounds helpful.